### ISSUE TITLE
Support fetching batches of Pub/Sub messages in SI channel adapters

### ISSUE THREAD

**Issue by meltsufin** [spring-cloud/spring-cloud-gcp#1962]
Currently our channel adapters only support working with individual messages, but significant performance gains can be achieved by supporting batches of messages.

**Comment by meltsufin** [comment]
More context from customer:

> At some point in our use case, we had to read a batch (list) of messages and process them at once and it felt strange missing that functionality. In the end, we had to implement this requirement with Spring Integration's Aggregator. This added a lot of implementation overhead even though we need only a list of messages.  
> 
> [We would like a] simpler way to pass a batch of messages to the `@ServiceActivator`. It is not a show stopper but It will be useful to have a better way of fetching batches.

**Comment by meltsufin** [comment]
I think  that using the SI Aggregator is the right approach for this use case.
However, we can make it easier by making sure that when we pull a batch using `PubSubMessageSource`, we add a batch ID header that can be used in a correlation strategy to reconstruct that batch.

@artembilan WDYT?

**Comment by artembilan** [comment]
I think the story is indeed about a pulling batch of messages exactly from the Pub/Sub. 
An aggregator just simulate a behavior and if there is really no batch support in Pub/Sub, there is nothing we can do unless simulate it similar way on the consumer level in the framework.
But doesn't it worth that if we can have the same behavior with already existing components?
Is it really a Pub/Sub Inbound Channel Adapter responsibility to buffer messages for end-user requirements?

**Comment by elefeint** [comment]
`PubSubMessageSource` already [gets baches](https://github.com/spring-cloud/spring-cloud-gcp/blob/480747b2b4e575e55e5a82292bd3ea4745c0556c/spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/integration/inbound/PubSubMessageSource.java#L103), though. But because `AbstractFetchLimitingMessageSource` contract is to receive one object at a time, we cache the rest.

We could implement a `PubSubBatchMessageSource` to return a list of messages, in principle.

**Comment by meltsufin** [comment]
+1 to creating a `PubSubBatchMessageSource`.
@artembilan What are you thoughts on that?

**Comment by artembilan** [comment]
Yes, that could be a clean solution. My point though was about a `PubSubInboundChannelAdapter` which doesn't look like comes with a batch option possible.
I somehow believe that was a request from end-user.
We need to keep in mind that pollable `MessageSource` is not always a solution for end-users...

**Comment by meltsufin** [comment]
`PubSubInboundChannelAdapter` doesn't because it's push-based (async).
`PubSubMessageSource` is our pull-based (sync) "inbound channel adapter".

**Comment by artembilan** [comment]
I know what you mean, but we need to be sure that we are on the same page with end-user.
So, if `BatchPubSubMessageSource` is good with them, we can proceed with an implementation.

No objections from my side!
